### ISSUE TITLE
Grace period on DeleteOptions has no effect on CRs

### ISSUE THREAD

**Issue by rmohr** [kubernetes/kubernetes#56567]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

We use a CR with a custom finalizer to get support for graceful deletes. While the crd apiserver seems to consider the deleteGracePeriod to some degree, it seems that it is not updating `metadata.deletionGracePeriodSeconds`. Therefore our finalizer has no way to differentiate between a forceful delete request `--grace-period 0` or a graceful delete request.

A follow up issue seems to be that `--grace-period` is not sent by kubectl if it talks to a cr. Therefore I tried it with a DeleteOptions object on a plain curl call too.

**What you expected to happen**:

I would expect that I can implement graceful delete based on a finalizer on my CR, in combination with sending `gracePeriodSecons` to the apiserver. This allows me to distinguish between a force delete and a graceful delete.

**How to reproduce it (as minimally and precisely as possible)**:

Here the kubectl log:

```bash
$ delete vms  testvm --grace-period 20 -v9
Request Body: {"orphanDependents":false}
I1129 12:13:33.298274     728 round_trippers.go:417] curl -k -v -XDELETE  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: .kubectl/v1.8.4 (linux/amd64) kubernetes/9befc2b" https://192.168.121.164:6443/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm
I1129 12:13:33.312421     728 round_trippers.go:436] DELETE https://192.168.121.164:6443/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm 202 Accepted in 14 milliseconds
I1129 12:13:33.312486     728 round_trippers.go:442] Response Headers:
I1129 12:13:33.312542     728 round_trippers.go:445]     Content-Type: application/json
I1129 12:13:33.312549     728 round_trippers.go:445]     Content-Length: 1198
I1129 12:13:33.312554     728 round_trippers.go:445]     Date: Wed, 29 Nov 2017 11:13:33 GMT
I1129 12:13:33.312621     728 request.go:836] Response Body: {"apiVersion":"kubevirt.io/v1alpha1","kind":"VirtualMachine","metadata":{"clusterName":"","creationTimestamp":"2017-11-29T11:07:54Z","deletionGracePeriodSeconds":0, [...]}
```
As you can see, no `gracePerodSeconds` is sent to the apiserver and the resulting delete response has `deletionGracePeriodSecons` set to 0.

And here the plain curl call:

```bash
$ curl -k -v -XDELETE  -H "Content-Type: application/json" -H "Accept: application/json" -H "User-Agent: .kubectl/v1.8.4 (linux/amd64) kubernetes/9befc2b" http://192.168.200.2:8184/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm --data '{"orphanDependents":false, "gracePeriodSeconds": 20}'
{"metadata":{"name":"testvm","namespace":"default","selfLink":"/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm","uid":"b7e81365-d4f6-11e7-8ec0-525400ea43c6","resourceVersion":"54319","creationTimestamp":"2017-11-29T11:16:17Z","deletionTimestamp":"2017-11-29T11:16:24Z","deletionGracePeriodSeconds":0, [...]}
```
This time I made sure by hand, that `gracePeriodSecons` is sent to the apiserver, but the resulting object still show 0 for `deletionGracePeriodSeconds`.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:28:34Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by nikhita** [comment]
From the example in the issue description, I think the issue is not being able to delete the "CustomResource" (not the CRD) gracefully.

However, I don't think we support graceful deletion for CRD as well...

This can probably done using:

https://github.com/kubernetes/kubernetes/blob/2aeace402ac8dc88440eb5c0453b31dd28ca14f2/staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go#L54-L60

/cc @sttts @deads2k wanted to double check if this is the way we want to go before working on a patch.

**Comment by rmohr** [comment]
@nikhita ah yes thx. I meant the Custom Resource (CR), not the CRD.

**Comment by enisoc** [comment]
@nikhita Is the plan to enable graceful delete on all CRD-backed CRs (and potentially also CRD itself)?

I'm not familiar with the implications of that, but I noticed that currently Pod is the only strategy in core that seems to implement `RESTGracefulDeleteStrategy` (i.e. it has a `CheckGracefulDelete` method). Maybe it should be opt-in on a per-CRD basis?

I'm also curious how the overall flow would work for a CR that supports graceful delete; are they responsible for having a controller come along and delete with grace-period=0 if the cleanup even they were waiting for happens before the original grace period expires? It seems strange to me that it would require both a finalizer and grace-period-seconds, since I don't think Pod uses a finalizer to achieve this.

In other words, if grace-period-seconds worked for CRs, I don't think you'd need a finalizer at all to make the OP's use case work.

**Comment by sttts** [comment]
Although github linked this already, here again explicitly: https://github.com/kubernetes/kubernetes/pull/60744

**Comment by enisoc** [comment]
@sttts Have you weighed in on whether that PR needs to be made opt-in with a field in the CRD? I'm assuming there is a reason graceful deletion is opt-in for native types (and only Pod seems to opt in), but I don't know the reason.

**Comment by rmohr** [comment]
> In other words, if grace-period-seconds worked for CRs, I don't think you'd need a finalizer at all to make the OP's use case work.

So the controller would then do a "force" delete once it is done, in contrast to removing the finalizer? This would then require the opt-in on the CRD, right? Instead  of removing the finalizer the controller would then do something like a `force` delete on the CR, once it thinks that it is ready for deletion? Would that be the equivalent of a `--force` from the `kubectl` point of view?

**Comment by sttts** [comment]
We have two options:
- set the `meta.gracefulDeletionPeriod` in any case on delete. A finalizer can then use it and only delete itself after this period.
- add a CRD `spec.gracefulDeletion bool` and implement real graceful deletion. In that case, the normal non-force delete request will only set the deletion timestamp. Another controller has to do a final forced deletion. The question is where such a controller should sit? In apiextensions-apiserver? In a 3rdparty controller?

**Comment by sttts** [comment]
I tend to the second option. @deads2k @liggitt @enisoc @caesarxuchao opinions?

**Comment by rmohr** [comment]
@sttts Created a POC for the second option in #63162.

**Comment by caesarxuchao** [comment]
I prefer the second option based on these comparisons:
* with option 1, both `meta.gracefulDeletePeriod` and a finalizer need to be set. option 2 doesn't require finalizer.
* with option 1, different CR implementers will use different finalizers to implement graceful deletion.
* with option 2, the flow is the same as the pod's graceful deletion.

**Comment by lavalamp** [comment]
We talked about this during the API Machinery SIG call, and the people there were mostly leaning towards recommending finalizers.
* Is graceful deletion a feature for any resource other than Pod? I'm not sure if we want to offer it generally.

**Comment by rmohr** [comment]
> Is graceful deletion a feature for any resource other than Pod? I'm not sure if we want to offer it generally.

It is definitely useful whenever you create a CR which represents resources on nodes and is managed there via daemons on these nodes.

When using finalizers only, the difference to pod graceful delete would be that a force delete requires a cluster-component of the extension to be present to "emulate" a pure force-delete by removing the finalizer. So if `gracefulDeletePeriodSeconds` is set to `0` the cluster-controller would immediately remove the finalizer which exists solely for emulating graceful deletes. Also `kubectl delete --force` would then work as expected.

**Comment by sttts** [comment]
Here is a PR for the first option: https://github.com/kubernetes/kubernetes/pull/63915. **It is ugly, our registry code is not made for that.**

With the second option (the one @rmohr implemented in #63162), we get a far simpler implementation that is in-line with the machinery we have. Implementation-wise I prefer that a lot.

@deads2k @lavalamp @cheftako @liggitt please take a look at both approaches. Do we really want to a finalizer based graceful deletion mechanisms like that? I would like to avoid that complexity, especially as we didn't have a strong and clear opinion during the Sig call about which route to go.

**Comment by akalenyu** [comment]
@rmohr this becomes an issue with https://github.com/kubevirt/kubevirt/pull/10383, unless this is fixed I think we can reopen

**Comment by dasionov** [comment]
The attemp to fix in #63915, is a bit too much IMO, all we need is just 
this `objectMeta.SetDeletionGracePeriodSeconds(options.GracePeriodSeconds)` no need to enforce the deletion 
what do you guys think? @rmohr @sttts @lavalamp @enisoc

**Comment by iholder101** [comment]
> /reopen

re-reopen
/reopen
